restler/engine/core/request_utilities.py

```diff
@@ -50,56 +53,95 @@ def str_to_hex_def(val_str):
+            elif token_auth_method == "module":
+                result = execute_token_refresh_module(token_dict["token_module_file"], token_dict["token_module_method"], token_dict["token_module_data"], token_dict["token_module_logging_enabled"])
+
+            _, latest_token_value, latest_shadow_token_value = parse_authentication_tokens(result)
             break
-        except subprocess.CalledProcessError:
-            error_str = f"Authentication failed when refreshing token:\n\nCommand that failed: \n{cmd}"
-            print(f'\n{error_str}')
-            latest_token_value = ERROR_VAL_STR
-            latest_shadow_token_value = ERROR_VAL_STR
-            _RAW_LOGGING(error_str)
-            retry_count = retry_count + 1
-            time.sleep(RETRY_SLEEP_TIME_SEC)
         except EmptyTokenException:
             error_str = "Error: Authentication token was empty."
             print(error_str)
             _RAW_LOGGING(error_str)
             sys.exit(-1)
         except Exception as error:
-            error_str = f"Exception refreshing token with cmd {cmd}.  Error: {error}"
-            print(error_str)
+            error_str = f"Authentication failed when refreshing token:\n\nUsing Token authentication method: \n{token_auth_method} \n with error {error}"
+            print(f'\n{error_str}')
+            latest_token_value = ERROR_VAL_STR
+            latest_shadow_token_value = ERROR_VAL_STR
             _RAW_LOGGING(error_str)
-            sys.exit(-1)
-    else:
-        _RAW_LOGGING(f"\nMaximum number of retries ({MAX_RETRIES}) exceeded. Exiting program.")
-        sys.exit(-1)
+            retry_handler.wait_for_next_retry()
+
+def execute_location_token_refresh(location):
+    try:
+        with open(location,"r") as f:
+            token_result = f.read()  
+            return token_result
+    except FileNotFoundError:
+        error_str = f"Could not find token file at {location}. Please ensure that you've passed a valid path"
+        _RAW_LOGGING(error_str)
+        raise EmptyTokenException(error_str)
+
+def execute_token_refresh_module(module_path, method, data, logging_enabled):
+    try:
+        module_name = os.path.basename(module_path)
+        spec = importlib.util.spec_from_file_location(module_name, module_path)
+        module = importlib.util.module_from_spec(spec)
+        sys.modules[module_name] = module
+        spec.loader.exec_module(module)
+    except FileNotFoundError: 
+        error_str = f"Could not find token module file at {module_path}/{module_name}. Please ensure that you've passed a valid path"
+        _RAW_LOGGING(error_str)
+        raise EmptyTokenException(error_str)
+    except AttributeError:
+        error_str = f"Could not load token module file at {module_path}/{module_name}. Please ensure that you've passed a valid python module"
+        _RAW_LOGGING(error_str)
+        raise EmptyTokenException(error_str)
```

Comment:
>EmptyTokenException [](http://example.com/codeflow?start=14&length=19)

Given the usage, how about naming this a more generic 'TokenRefreshException'?

Reply:
We are retrying generic exceptions in the parent right? Do you think we should fail fast in case of FileNotFoundException and AttributeError? I was thinking that retries probably won't help in these cases, but it could be useful to retry other generic exceptions depending on consumers specific auth method/implementation

Reply:
Ah, good point.  In that case, could you please define another type of exception and fail fast in the caller for these types of errors?  I found this a bit misleading since it is not an empty token, but an error in user-specified settings.